swap/swap.go

```diff
@@ -1,98 +1,451 @@
+	// that the balance is *below* the threshold
+	if newBalance <= -s.paymentThreshold {
+		//if so, send cheque
+		log.Warn("balance for peer went over the payment threshold, sending cheque", "peer", peer.ID().String(), "payment threshold", s.paymentThreshold)
+		return s.sendCheque(peer.ID())
+	}
+
+	return nil
+}
+
+func (s *Swap) updateBalance(peer enode.ID, amount int64) (int64, error) {
 	//adjust the balance
 	//if amount is negative, it will decrease, otherwise increase
-	s.balances[peer.ID()] += amount
+	s.balances[peer] += amount
 	//save the new balance to the state store
-	peerBalance := s.balances[peer.ID()]
-	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	peerBalance := s.balances[peer]
+	err := s.store.Put(balanceKey(peer), &peerBalance)
+	if err != nil {
+		return 0, fmt.Errorf("error while storing balance for peer %s", peer.String())
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(peerBalance, 10))
+	return peerBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
-	return err
+// loadBalance loads balances from the state store (persisted)
+func (s *Swap) loadBalance(peer enode.ID) (err error) {
+	var peerBalance int64
+	//only load if the current instance doesn't already have this peer's
+	//balance in memory
+	if _, ok := s.balances[peer]; !ok {
+		err = s.store.Get(balanceKey(peer), &peerBalance)
+		s.balances[peer] = peerBalance
+	}
+	return
 }
 
-//GetPeerBalance returns the balance for a given peer
-func (swap *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
-	swap.lock.RLock()
-	defer swap.lock.RUnlock()
-	if p, ok := swap.balances[peer]; ok {
-		return p, nil
+// sendCheque sends a cheque to peer
+func (s *Swap) sendCheque(peer enode.ID) error {
+	swapPeer, err := s.getPeer(peer)
+	if err != nil {
+		return fmt.Errorf("error while getting peer: %s", err.Error())
+	}
+	cheque, err := s.createCheque(peer)
+	if err != nil {
+		return fmt.Errorf("error while creating cheque: %s", err.Error())
+	}
+
+	log.Info("sending cheque", "serial", cheque.ChequeParams.Serial, "amount", cheque.ChequeParams.Amount, "beneficiary", cheque.Beneficiary, "contract", cheque.Contract)
+	s.cheques[peer] = cheque
+
+	err = s.store.Put(sentChequeKey(peer), &cheque)
+	if err != nil {
+		return fmt.Errorf("error while storing the last cheque: %s", err.Error())
+	}
+
+	emit := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+
+	// reset balance;
+	err = s.resetBalance(peer, int64(cheque.Amount))
+	if err != nil {
+		return err
 	}
-	return 0, errors.New("Peer not found")
+
+	return swapPeer.Send(context.Background(), emit)
 }
 
-//load balances from the state store (persisted)
-func (s *Swap) loadState(peer *protocols.Peer) (err error) {
-	var peerBalance int64
-	peerID := peer.ID()
+// Create a Cheque structure emitted to a specific peer as a beneficiary
```

Comment:
`Create a new cheque, whose serial and amount...`

Reply:
fixed